### PR TITLE
feat: enable serverless and rawdeployment deployment mode

### LINKED ISSUE #82: Ingress configuration is hardcoded
[ingress configuration](https://github.com/canonical/kserve-operators/blob/main/charms/kserve-controller/src/templates/configmap_manifests.yaml.j2#L48) is hardcoded, which expects ingress to be deployed in a certain way. We should change the way this is done.

### PR DESCRIPTION
**Pull request by DnPlas**
This commit enables different charm configurations and relations to be able to set up the deployment mode.

Fixes #82 

TODO:
- [x] add unit tests
- [x] add integration tests
- [x] gateway-info -> ingress-gateway relation name in metadata

### CODE REVIEW

File: charms/kserve-controller/src/charm.py [hunk 1]
```diff
@@ -76,6 +89,25 @@ def _context(self):
...
         }
 
+    @property
+    def _inference_service_context(self):
+        """Context for rendering the inferenceservive-config ConfigMap."""
+        # Ensure any input is valid for deployment mode
+        deployment_mode = self.model.config["deployment-mode"].lower()
+        if deployment_mode == "serverless":
+            deployment_mode = "Serverless"
+        deployment_mode = "RawDeployment"
```
Comment (beliaev-maksim) [inline charms/kserve-controller/src/charm.py:99]: I do not know, if the user makes a typo?

what if we go into blocked state if the option is incorrect?
```
if serverless:
elif raw:
else: Blocked
```

File: charms/kserve-controller/src/charm.py [hunk 2]
```diff
@@ -203,6 +261,73 @@ def _on_remove(self, _):
...
+            "ingress_gateway_namespace": ingress_gateway_info["gateway_namespace"],
+            "ingress_gateway_service_name": "istio-ingressgateway-workload",
+            "local_gateway_name": "",
+            "local_gateway_namespace": "",
+            "local_gateway_service_name": "",
+        }
+
+        # Get the local-gateway info. This value should only
+        # be get and rendered in Serverless Mode.
+        if self.model.config["deployment_mode"] == "Serverless":
```
Comment (beliaev-maksim) [inline charms/kserve-controller/src/charm.py:309]: up in the stack we are case insensitive

File: charms/kserve-controller/config.yaml [hunk 3]
```diff
@@ -1,6 +1,14 @@
 # Copyright 2022 Canonical Ltd.
 # See LICENSE file for licensing details.
 options:
+  deployment-mode:
+    default: "RawDeployment"
+    description: Deployment mode for kserve. It can only be RawDeployment or Serverless, and the latter requires knative-serving to be present in the model, and the local-gateway relation to be established.
+    type: str
+  domain-name:
```
Comment (beliaev-maksim) [inline charms/kserve-controller/config.yaml:8]: can we check in the code that it is the same as in knative ?

File: charms/kserve-controller/src/charm.py (hunk 2 above)
Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:309]: true, good catch

File: charms/kserve-controller/src/charm.py (hunk 1 above)
Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:99]: yeah, I can add something.

File: charms/kserve-controller/config.yaml (hunk 3 above)
Comment (DnPlas) [reply charms/kserve-controller/config.yaml:8]: It's a bit hard because knative-serving is not sharing that data, for that I would need to share it somehow and that would require to either modify the gateway info library or create a new one just for knative.

Comment (beliaev-maksim) [reply charms/kserve-controller/config.yaml:8]: let's treat it as the enhancement which is out of this scope. Can you open an issue to track it?

Comment (DnPlas) [reply charms/kserve-controller/config.yaml:8]: Tracked in #98

File: charms/kserve-controller/src/charm.py [hunk 4]
```diff
@@ -36,28 +41,36 @@
     "src/templates/auth_manifests.yaml.j2",
     "src/templates/serving_runtimes_manifests.yaml.j2",
     "src/templates/webhook_manifests.yaml.j2",
-    "src/templates/configmap_manifests.yaml.j2",
 ]
-CONFIG_FILE = "src/config/inference-config.j2"
+CONFIG_FILE = ["src/templates/configmap_manifests.yaml.j2"]
```
Comment (beliaev-maksim) [inline charms/kserve-controller/src/charm.py:45]:
```suggestion
CONFIG_FILES = ["src/templates/configmap_manifests.yaml.j2"]
```

make plural? :)

File: charms/kserve-controller/metadata.yaml [hunk 5]
```diff
@@ -8,13 +8,11 @@ description: |
   KServe Controller.
 summary: |
   KServe Controller.
-
```
Comment (i-chvets) [inline charms/kserve-controller/metadata.yaml:11]: Do we want to add doc links here or should we do it in separate PR?

File: charms/kserve-controller/src/charm.py [hunk 6]
```diff
@@ -185,10 +244,19 @@ def _on_install(self, event):
...
             self.k8s_resource_handler.apply()
-        except ApiError as e:
-            log.error(e)
+            self.cm_resource_handler.apply()
+        except ErrorWithStatus as err:
+            self.model.unit.status = err.status
+            log.error(f"Failed to handle {event} with error: {err}")
+            return
+        except ApiError as api_err:
+            log.error(api_err)
```
Comment (i-chvets) [inline charms/kserve-controller/src/charm.py:253]: At this point we will stay in MaintenanceStatus, correct?
Should we raise error here as well?

File: charms/kserve-controller/src/charm.py [hunk 7]
```diff
@@ -36,28 +41,36 @@
...
             self.on.kube_rbac_proxy_pebble_ready, self._on_kube_rbac_proxy_ready
         )
-
+        self.framework.observe(
+            self.on["ingress-gateway"].relation_changed, self._on_ingress_gateway_relation_changed
+        )
+        self.framework.observe(
+            self.on["local-gateway"].relation_changed, self._on_local_gateway_relation_changed
+        )
         self._k8s_resource_handler = None
```
Comment (i-chvets) [inline charms/kserve-controller/src/charm.py:71]: When relation is removed (broken or departed) should we have a special handling that puts charm back into Blocked state?

File: charms/kserve-controller/metadata.yaml (hunk 5 above)
Comment (DnPlas) [reply charms/kserve-controller/metadata.yaml:11]: Good catch, but we don't have the kserve docs in [the bundle](https://charmed-kubeflow.io/docs/kubeflow-bundle). I'll submit a new patch for that.

File: charms/kserve-controller/src/charm.py (hunk 7 above)
Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:71]: Yeah, good point.

Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:79]: Updated it, please let me know what you think.

File: charms/kserve-controller/src/charm.py [hunk 8]
```diff
@@ -76,6 +97,30 @@ def _context(self):
...
+        # Ensure any input is valid for deployment mode
+        deployment_mode = self.model.config["deployment-mode"].lower()
+        if deployment_mode == "serverless":
+            deployment_mode = "Serverless"
+        elif deployment_mode == "rawdeployment":
+            deployment_mode = "RawDeployment"
+        else:
+            self.unit.status = BlockedStatus(
+                "Please set deployment-mode to either Serverless or RawDeployment"
+            )
```
Comment (beliaev-maksim) [inline charms/kserve-controller/src/charm.py:112]: should not we exit from here?

otherwise we proceed with the wrong depl mode

**Review by beliaev-maksim — DISMISSED** [review]

File: charms/kserve-controller/src/charm.py (hunk 7 above)
Comment (i-chvets) [reply charms/kserve-controller/src/charm.py:79]: Perfect!

File: charms/kserve-controller/metadata.yaml (hunk 5 above)
Comment (i-chvets) [reply charms/kserve-controller/metadata.yaml:11]: Great!

**Comment by i-chvets** [comment]
LGTM.

**Review by i-chvets — APPROVED** [review]
Approved.

File: charms/kserve-controller/src/charm.py (hunk 8 above)
Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:112]: I have updated this with a `return`. Let me know what you think.

File: charms/kserve-controller/src/charm.py [hunk 9]
```diff
@@ -76,6 +97,31 @@ def _context(self):
...
+            return
```
Comment (beliaev-maksim) [inline charms/kserve-controller/src/charm.py:113]: but this will return None to `cm_resource_handler`

and will proceed code execution. I think it is not what we expected, or?

Comment (DnPlas) [reply charms/kserve-controller/src/charm.py:113]: Would it be better to raise ErrorWithStatus here? It will be handled in `_on_event()`

Comment (beliaev-maksim) [reply charms/kserve-controller/src/charm.py:113]: looks like yes, the only way

or 1 million returns :D

File: charms/kserve-controller/tests/integration/test_charm.py [hunk 10]
```diff
@@ -16,6 +16,9 @@
 
 logger = logging.getLogger(__name__)
 
+ISTIO_VERSION = "1.16/stable"
```
Comment (i-chvets) [inline charms/kserve-controller/tests/integration/test_charm.py:19]: @DnPlas Do we want to use `latest/edge` here?
There were updates to Istio that might solve integration test problems.

Comment (DnPlas) [reply charms/kserve-controller/tests/integration/test_charm.py:19]: I tried latest/edge first, then changed it to 1.16/stable to see if that would fix the errors, but it did not.

**Review by beliaev-maksim — APPROVED** [review]

**Review by i-chvets — APPROVED** [review]
Approved.
